dask/dataframe/core.py

```diff
@@ -4672,15 +4673,82 @@ def set_index(
         >>> # ^ Now copy-paste this and edit the line above to:
         >>> # ddf2 = ddf.set_index("name", divisions=["Alice", "Laura", "Ursula", "Zelda"])
         """
+
+        # Deal with bool params
         if inplace:
             raise NotImplementedError("The inplace= keyword is not supported")
         pre_sorted = sorted
         del sorted
 
-        if isinstance(other, Series) and other._name == self.index._name:
-            # Index is equal to current index, no-op
-            return self
+        # Check other can be translated to column name or column object, possibly flattening it
+        if not isinstance(other, str):
+            if isinstance(other, Sequence):
+                # Accept ["a"], but not [["a"]]
+                if len(other) == 1 and (
+                    isinstance(other[0], str) or not isinstance(other[0], Sequence)
+                ):
+                    other = other[0]
+                else:
+                    raise NotImplementedError(
+                        "Dask dataframe does not yet support multi-indexes.\n"
+                        f"You tried to index with this index: {other}\n"
+                        "Indexes must be single columns only."
+                    )
+
+            # Can also be a frame
+            elif isinstance(other, DataFrame):
+                raise NotImplementedError(
+                    "Dask dataframe does not yet support multi-indexes.\n"
+                    f"You tried to index with a frame with these columns: {list(other.columns)}\n"
+                    "Indexes must be single columns only."
+                )
 
+        # If already a series
+        index, columns = self.index, self.columns
+        if is_series := isinstance(other, Series):
+            # If it's already the index, there's nothing to do
+            if other._name == index._name:
+                return self
+            # Otherwise, check length matches when other isn't one of the data columns
+            s = other
+            if not any(
+                s._name == col_series._name for _, col_series in self.items()
+            ) and (  # for cases with other=df.column
+                len_index := len(s)
+            ) != (
+                len_self := len(self)
+            ):
```

Comment:
No, cannot do that check because those in the columns are `.name`, but not `._name`, we need to check the actual `Series` instances. Anyway, I changed the check to the more clear:

```python
if not any(s._name == self[c]._name for c in self.columns)
```